s3utils/path.go

```diff
@@ -1,24 +1,47 @@
 package s3utils
 
 import (
+	"fmt"
 	"net/url"
-	"os"
+	"path"
 
-	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/s3"
 )
 
-// S3PathToBucketAndKey converts an s3 path into it's bucket and key
-func S3PathToBucketAndKey(s3path string) (string, string, error) {
+// Path is an interface of functions to apply transparently to s3 or local paths
+type Path interface {
+	IsDir() (bool, error)
+	Exists() (bool, error)
+	IsS3() bool
+	IsLocal() bool
+	ListPathsWithPrefix() ([]Path, error)
+	Join(...string) Path
+	Base() string
+	WithoutPrefix(Path) string
+	ToStringWithoutBucket() string
+	Bucket() (string, error)
+	String() string
+}
+
+// s3PathToBucketAndKey converts an s3 path into it's bucket and key
+func s3PathToBucketAndKey(s3path string) (string, string, error) {
 	url, err := url.Parse(s3path)
 	if err != nil {
 		return "", "", err
 	}
-	return url.Host, url.Path[1:], nil
+	key := ""
+	if url.Path != "/" && url.Path != "" {
+		key = url.Path[1:]
+	}
+	return url.Host, key, nil
+}
+
+func bucketAndKeyToS3Path(bucket string, key string) string {
+	return fmt.Sprintf("s3://%s", path.Join(bucket, key))
 }
 
-// IsS3Path checks whether a string is an s3 path
-func IsS3Path(path string) bool {
+// isS3Path checks whether a string is an s3 path
+func isS3Path(path string) bool {
```

Comment:
This capitalization looks like a bug b/c you're calling `s3utils.IsS3Path` elsewhere

Reply:
I didn't see any usages of `s3utils.IsS3Path`. The way this should be done is:

```go
path, _ := s3utils.NewPath(client, myPath)
path.IsS3()
```

`isS3Path` is a private function for internal use by `NewPath` only

Reply:
ah, I was looking at the deleted side 👍